### PR TITLE
Add RECURRENCE-ID with range parameter.

### PR DESCRIPTION
**Pull request by jabiertxof**
Add RECURRENCE-ID with range parameter.

contribute to https://github.com/niccokunzmann/python-recurring-ical-events/issues/75

### CODE REVIEW

**Comment by jabiertxof** [comment]
WIP: Need to include rrule

**Comment by jabiertxof** [comment]
Added test for rdate

File: test/conftest.py [hunk 1]
```diff
@@ -1,8 +1,13 @@
...
+import sys,os
 import time
 from datetime import timezone
 from pathlib import Path
-
+try:
+    DESKTOPPATH = Path(__file__).resolve().parent
+except:
+    print("error encontrando ruta de la aplicacione")
+sys.path.append(os.path.join(Path(__file__).resolve().parent, 'inc'))
```
Comment (niccokunzmann) [inline test/conftest.py:9]: I can send a refactoring for this but it is alright.
With Path() you do `Path(__file__) / 'inc'`

**Review by niccokunzmann — CHANGES_REQUESTED** [review]
Hi, I request changes but first of all! Cool! This looks really good! Let me know which changes you would like to tackle and which ones not so I can create a PR to your branch for those that you do not like ;)

Also, to close the issue we need more tests, I believe..

File: recurring_ical_events.py [hunk 2]
```diff
@@ -720,6 +735,18 @@ def recurrence_ids(self) -> RecurrenceIDs:
             return ()
         return to_recurrence_ids(recurrence_id.dt)
 
+    @cached_property
+    def thisandfuture(self) -> bool:
```
[the remark this answers is not in the document]
Comment (jabiertxof) [reply recurring_ical_events.py:739]: thanks

File: recurring_ical_events.py [hunk 3]
```diff
@@ -650,7 +665,7 @@ def __repr__(self):
 class ComponentAdapter(ABC):
     """A unified interface to work with icalendar components."""
 
-    ATTRIBUTES_TO_DELETE_ON_COPY = ["RRULE", "RDATE", "EXDATE"]
```
[the remark this answers is not in the document]
Comment (jabiertxof) [reply recurring_ical_events.py:653]: fixed

File: recurring_ical_events.py [hunk 4]
```diff
@@ -606,7 +608,18 @@ def between(self, span_start: Time, span_stop: Time) -> Generator[Occurrence]:
...
-            if adapter is self.recurrence.core:
+            if prev_adapter and adapter is self.recurrence.core:
+                start_dt = datetime.datetime.combine(start.date(), prev_adapter.start.time())
+                stop = get_any(
+                    self.recurrence.replace_ends,
+                    recurrence_ids,
+                    normalize_pytz(start_dt + prev_adapter.duration),
+                )
+                core = self.recurrence.core
+                self.recurrence.core = prev_adapter
```
[the remark this answers is not in the document]
Comment (jabiertxof) [reply recurring_ical_events.py:619]: fixed

File: recurring_ical_events.py [hunk 5]
```diff
@@ -606,7 +608,18 @@ def between(self, span_start: Time, span_stop: Time) -> Generator[Occurrence]:
             adapter: ComponentAdapter = get_any(
                 self.recurrence_id_to_modification, recurrence_ids, self.recurrence.core
             )
-            if adapter is self.recurrence.core:
+            if prev_adapter and adapter is self.recurrence.core:
```
[the remark this answers is not in the document]
Comment (jabiertxof) [reply recurring_ical_events.py:611]: thanks

File: test/conftest.py [hunk 6]
```diff
@@ -1,8 +1,13 @@
...
+print(os.path.join(Path(__file__).resolve().parent, 'inc'))
```
[the remark this answers is not in the document]
Comment (jabiertxof) [reply test/conftest.py:10]: legacy code to test local removed

File: test/test_issue_75_range_parameter.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    ],
+)
+def test_issue_75_RANGE_parameter(calendars, date, summary):
+    events = calendars.issue_75_range_parameter.at(date)
+    assert len(events) == 1, f"Expecting one event at {date}"
+    event = events[0]
+    print( str(event["SUMMARY"]),date,summary)
+    assert str(event["SUMMARY"]) == summary
+
+# TODO: Test DTSTART and DTEND
```
[the remark this answers is not in the document]
Comment (jabiertxof) [reply test/test_issue_75_range_parameter.py:68]: done

File: recurring_ical_events.py [hunk 8]
```diff
@@ -595,7 +596,8 @@ def between(self, span_start: Time, span_stop: Time) -> Generator[Occurrence]:
         returned_modifications: set[ComponentAdapter] = set()
         # NOTE: If in the following line, we get an error, datetime and date
         # may still be mixed because RDATE, EXDATE, start and rule.
-        for start in self.recurrence.rrule_between(span_start, span_stop):
+        prev_adapter = None
```
[the remark this answers is not in the document]
Comment (jabiertxof) [reply recurring_ical_events.py:599]: done

File: recurring_ical_events.py [hunk 9]
```diff
@@ -519,7 +520,7 @@ def rrule_between(self, span_start: Time, span_stop: Time) -> Generator[Time]:
                     span_stop_dt + datetime.timedelta(hours=1)
                 )
             for rule in self.rrules:
-                for start in rule.between(span_start_dt, span_stop_dt, inc=True):
+                for start in rule.between(rule._dtstart, span_stop_dt, inc=True):
```
[the remark this answers is not in the document]
Comment (jabiertxof) [reply recurring_ical_events.py]: fixed

**Comment by niccokunzmann** [comment]
Thanks for your additions! I will merge this and open another PR with the additional tests.. This does not yet solve the issue but it is really close.

**Pull request merged**